core/src/main/java/org/jruby/truffle/nodes/control/DoWhileNode.java

```diff
@@ -67,6 +65,7 @@ public boolean executeRepeating(VirtualFrame frame) {
                     break;
                 } catch (RedoException e) {
                     // Just continue in the while(true) loop.
+                    context.getSafepointManager().poll();
```

Comment:
Yes `redo` is unusual - most loops will run this inner while loop just once per actual loop iteration. I think we need to poll the safepoint before the return as well. This duplication of safepoints might be why I put it at the head in the first place.